include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -525,337 +544,790 @@ namespace internal
-                    VectorizedArray<Number> *scratch_data,
+    void integrate (const MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
+                    Number                  *values_dofs,
+                    Number                  *values_quad,
+                    Number                  *gradients_quad,
+                    Number                  *scratch_data,
                     const bool               integrate_values,
-                    const bool               integrate_gradients);
+                    const bool               integrate_gradients,
+                    const bool               overwrite_values);
   };
 
-  template <int dim, int fe_degree, int n_components, typename Number>
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
   inline
   void
-  FEEvaluationImplCollocation<dim, fe_degree, n_components, Number>
-  ::evaluate (const MatrixFreeFunctions::ShapeInfo<VectorizedArray<Number>> &shape_info,
-              VectorizedArray<Number> *values_dofs[],
-              VectorizedArray<Number> *values_quad[],
-              VectorizedArray<Number> *gradients_quad[][dim],
-              VectorizedArray<Number> *hessians_quad[][(dim*(dim+1))/2],
-              VectorizedArray<Number> *,
-              const bool               evaluate_values,
-              const bool               evaluate_gradients,
-              const bool               evaluate_hessians)
+  FEEvaluationImplTransformToCollocation<dim, fe_degree, n_q_points_1d, n_components, Number>
+  ::evaluate (const MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
+              const Number                                 *values_dofs,
+              Number                                       *values_quad,
+              Number                                       *gradients_quad,
+              Number                                       *hessians_quad,
+              Number *,
+              const bool                                    ,
+              const bool                                    evaluate_gradients,
+              const bool                                    evaluate_hessians)
   {
-    typedef EvaluatorTensorProduct<evaluate_evenodd, dim, fe_degree, fe_degree+1,
-            VectorizedArray<Number> > Eval;
-    Eval eval(AlignedVector<VectorizedArray<Number> >(),
-              shape_info.shape_gradients_eo,
-              shape_info.shape_hessians_eo,
-              shape_info.fe_degree,
-              shape_info.n_q_points_1d);
-
-    // These avoid compiler warnings; they are only used in sensible context
-    // but compilers typically cannot detect when we access something like
-    // gradients_quad[2] only for dim==3.
-    const unsigned int d1 = dim>1?1:0;
-    const unsigned int d2 = dim>2?2:d1;
-    const unsigned int d3 = d1+d2;
-    const unsigned int d4 = dim>2?4:d3;
-    const unsigned int d5 = dim>2?5:d4;
+    Assert(n_q_points_1d > fe_degree,
```

Comment:
The compiler does not like the static assert because the function where we select the different paths sees the struct also in the invalid case - even though the code then does not enter the function.